### The Libraries

This part of the project is where I will be importing all of the libraries I will use throughout the project.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import math

### The Data

This part is intended to just upload all of the data that will be used for the project as dataframes.

In [2]:
fi_2014 = pd.read_csv("Financial_Indicators/2014_Financial_Data.csv")
fi_2015 = pd.read_csv("Financial_Indicators/2015_Financial_Data.csv")
fi_2016 = pd.read_csv("Financial_Indicators/2016_Financial_Data.csv")
fi_2017 = pd.read_csv("Financial_Indicators/2017_Financial_Data.csv")
fi_2018 = pd.read_csv("Financial_Indicators/2018_Financial_Data.csv")

In [3]:
for i in fi_2014.columns:
    print(i)

Unnamed: 0
Revenue
Revenue Growth
Cost of Revenue
Gross Profit
R&D Expenses
SG&A Expense
Operating Expenses
Operating Income
Interest Expense
Earnings before Tax
Income Tax Expense
Net Income - Non-Controlling int
Net Income - Discontinued ops
Net Income
Preferred Dividends
Net Income Com
EPS
EPS Diluted
Weighted Average Shs Out
Weighted Average Shs Out (Dil)
Dividend per Share
Gross Margin
EBITDA Margin
EBIT Margin
Profit Margin
Free Cash Flow margin
EBITDA
EBIT
Consolidated Income
Earnings Before Tax Margin
Net Profit Margin
Cash and cash equivalents
Short-term investments
Cash and short-term investments
Receivables
Inventories
Total current assets
Property, Plant & Equipment Net
Goodwill and Intangible Assets
Long-term investments
Tax assets
Total non-current assets
Total assets
Payables
Short-term debt
Total current liabilities
Long-term debt
Total debt
Deferred revenue
Tax Liabilities
Deposit Liabilities
Total non-current liabilities
Total liabilities
Other comprehensive income
Re

In [4]:
# This is a quick check of one of the files.  The files should have very similar information
# with the only difference between them being the year the data was collected.

fi_2014.head()

Unnamed: 0       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0         PG  7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   
1       VIPS  3.734148e+09          1.1737     2.805625e+09  9.285226e+08   
2         KR  9.837500e+10          0.0182     7.813800e+10  2.023700e+10   
3        RAD  2.552641e+10          0.0053     1.820268e+10  7.323734e+09   
4        GIS  1.790960e+10          0.0076     1.153980e+10  6.369800e+09   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0  0.000000e+00  2.146100e+10        2.146100e+10      1.391000e+10   
1  1.083303e+08  3.441414e+08        7.939267e+08      1.345959e+08   
2  0.000000e+00  1.519600e+10        1.751200e+10      2.725000e+09   
3  0.000000e+00  6.561162e+09        6.586482e+09      7.372520e+08   
4  0.000000e+00  3.474300e+09        3.412400e+09      2.957400e+09   

   Interest Expense  ...  Receivables growth  Inventory Growth  Asset Growth  \
0      7.090000e+08  ...             -0.0187           -0.0217        0.0359   
1      1.214869e+07  ...                 NaN               NaN           NaN   
2      4.430000e+08  ...              0.0618            0.0981        0.1886   
3      4.245910e+08  ...              0.0211           -0.0510       -0.0189   
4      3.024000e+08  ...              0.0257            0.0090        0.0215   

   Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                       0.0316       0.1228              0.0000   
1                          NaN          NaN              1.6484   
2                       0.3268       0.2738              0.0000   
3                       0.1963      -0.0458              0.0000   
4                       0.0274       0.1025              0.0000   

   SG&A Expenses Growth              Sector  2015 PRICE VAR [%]  Class  
0               -0.1746  Consumer Defensive           -9.323276      0  
1                1.7313  Consumer Defensive          -25.512193      0  
2                0.0234  Consumer Defensive           33.118297      1  
3               -0.0060  Consumer Defensive            2.752291      1  
4               -0.0220  Consumer Defensive           12.897715      1  

[5 rows x 225 columns]

Now I will chekc the shape of all those dataframes.

In [5]:
# This is a quick check on the shape of the datasets.

print('The shape of the 2014 file is:', fi_2014.shape)

print('The shape of the 2015 file is:', fi_2015.shape)

print('The shape of the 2016 file is:', fi_2016.shape)

print('The shape of the 2017 file is:', fi_2017.shape)

print('The shape of the 2018 file is:', fi_2018.shape)

The shape of the 2014 file is: (3808, 225)
The shape of the 2015 file is: (4120, 225)
The shape of the 2016 file is: (4797, 225)
The shape of the 2017 file is: (4960, 225)
The shape of the 2018 file is: (4392, 225)


The shape of the files shows that each one of them have different amounts of rows, which means that some of the files will have more stocks than others.  

The good news is that they all have the same amount of columns/features.  This is good because it makes it possible to merge those files into a single dataframe which makes the project more effective.

The plan is to merge all of these dataframes into a single dataframe.  Since each dataframe is from a different year, before merging them, I will check the sets of stocks in each individual dataframe, which in this case are the very first column of each dataframe.  There needs to be only one of each in each year to avoid repeated information.  To do this check, I will confirm that the amount of unique stocks is equals to the number of rows for each dataframe.  True means the amount of rows and amount of stocks are the same.

In [6]:
print('The check for the 2014 dataframe is', len(fi_2014['Unnamed: 0'].unique()) == fi_2014.shape[0])

print('The check for the 2015 dataframe is', len(fi_2015['Unnamed: 0'].unique()) == fi_2015.shape[0])

print('The check for the 2016 dataframe is', len(fi_2016['Unnamed: 0'].unique()) == fi_2016.shape[0])

print('The check for the 2017 dataframe is', len(fi_2017['Unnamed: 0'].unique()) == fi_2017.shape[0])

print('The check for the 2018 dataframe is', len(fi_2018['Unnamed: 0'].unique()) == fi_2018.shape[0])

The check for the 2014 dataframe is True
The check for the 2015 dataframe is True
The check for the 2016 dataframe is True
The check for the 2017 dataframe is True
The check for the 2018 dataframe is True


Great!  Now we know there are not repeat stocks in each of the dataframes!!!

Now let's perform the same test for the columns.  Each set of unique columns will be compared to the amount of columns in each dataframe.

In [7]:
print('The check for the 2014 dataframe is', len(fi_2014.columns.unique()) == fi_2014.shape[1])

print('The check for the 2015 dataframe is', len(fi_2015.columns.unique()) == fi_2015.shape[1])

print('The check for the 2016 dataframe is', len(fi_2016.columns.unique()) == fi_2016.shape[1])

print('The check for the 2017 dataframe is', len(fi_2017.columns.unique()) == fi_2017.shape[1])

print('The check for the 2018 dataframe is', len(fi_2018.columns.unique()) == fi_2018.shape[1])

The check for the 2014 dataframe is True
The check for the 2015 dataframe is True
The check for the 2016 dataframe is True
The check for the 2017 dataframe is True
The check for the 2018 dataframe is True


Perfect!

The financial indicators in the year 2014 is a good way we can see how the dataset is built.  Since we are dealing with multiple datasets that should have the same kind of information, the column names in all of them should be the same.  That means that the next step is to check just that.

In [8]:
# This step will isolate the column names to check if all files have equal columns.

col_2014 = fi_2014.columns # Isolating the 2014 column names

col_2015 = fi_2015.columns # Isolating the 2015 column names

col_2016 = fi_2016.columns # Isolating the 2016 column names

col_2017 = fi_2017.columns # Isolating the 2017 column names

col_2018 = fi_2018.columns # Isolating the 2018 column names

# Now let's print the column names for the 2014 file just to familiarize with them.
for f in col_2014:
    print(f)

Unnamed: 0
Revenue
Revenue Growth
Cost of Revenue
Gross Profit
R&D Expenses
SG&A Expense
Operating Expenses
Operating Income
Interest Expense
Earnings before Tax
Income Tax Expense
Net Income - Non-Controlling int
Net Income - Discontinued ops
Net Income
Preferred Dividends
Net Income Com
EPS
EPS Diluted
Weighted Average Shs Out
Weighted Average Shs Out (Dil)
Dividend per Share
Gross Margin
EBITDA Margin
EBIT Margin
Profit Margin
Free Cash Flow margin
EBITDA
EBIT
Consolidated Income
Earnings Before Tax Margin
Net Profit Margin
Cash and cash equivalents
Short-term investments
Cash and short-term investments
Receivables
Inventories
Total current assets
Property, Plant & Equipment Net
Goodwill and Intangible Assets
Long-term investments
Tax assets
Total non-current assets
Total assets
Payables
Short-term debt
Total current liabilities
Long-term debt
Total debt
Deferred revenue
Tax Liabilities
Deposit Liabilities
Total non-current liabilities
Total liabilities
Other comprehensive income
Re

In [9]:
# Let's check how the 2014 column names compares to the other files.

print('The number of unequal column names comparing the 2014 dataset with the others:')

print('2014 - 2015:', (col_2014 != col_2015).sum()) # Checking amount of unmatching columns

print('2014 - 2016:', (col_2014 != col_2016).sum()) # Checking amount of unmatching columns

print('2014 - 2017:', (col_2014 != col_2017).sum()) # Checking amount of unmatching columns

print('2014 - 2018:', (col_2014 != col_2018).sum()) # Checking amount of unmatching columns

The number of unequal column names comparing the 2014 dataset with the others:
2014 - 2015: 1
2014 - 2016: 1
2014 - 2017: 1
2014 - 2018: 1


In [10]:
# Let's check which column from the 2014 file does not match the other files column names.

print('The different column name is:\n')

print('2014 - 2015:\n', col_2014[col_2014 != col_2015][0]) # Checking which column name is not matching
print('\n')
print('2014 - 2016:\n', col_2014[col_2014 != col_2016][0]) # Checking which column name is not matching
print('\n')
print('2014 - 2017:\n', col_2014[col_2014 != col_2017][0]) # Checking which column name is not matching
print('\n')
print('2014 - 2018:\n', col_2014[col_2014 != col_2018][0]) # Checking which column name is not matching
print('\n')

The different column name is:

2014 - 2015:
 2015 PRICE VAR [%]


2014 - 2016:
 2015 PRICE VAR [%]


2014 - 2017:
 2015 PRICE VAR [%]


2014 - 2018:
 2015 PRICE VAR [%]




It seems that since the column title (2015 PRICE VAR [%]) contains a 'year' in it.  Since in 2014 the year said 2015, that means that the values in that column is taken at the end of the data year, which means at the beginning of the following year. It can be assumed that for the other years the name will be different, for example the 2015 data will have the column '2016 PRICE VAR [%]'.  What needs to be done is to change the column name by removing the year from it.

The other column name that I feel like needs to be changed is the very first column, which currently is 'Unnamed'.  The name of these columns in each of the dataset will become 'Stock'.

To separate the data by the year, I will create a new column named "Year" which will contain the integer value for the year of the dataset.  This column is created to help organize the dataset when merging them since I believe that many of the stock names will be the same throughout all of the datasets.

In [11]:
fi_2014.rename(columns = {'Unnamed: 0' : 'Stock', # Changing 'Stock' column name
                          '2015 PRICE VAR [%]' : 'PRICE VAR [%]'}, inplace = True) # Changing 'PRICE VAR' column name
fi_2014['Year'] = 2014 # Creating 'Year' Column

fi_2015.rename(columns = {'Unnamed: 0' : 'Stock', # Changing 'Stock' column name
                          '2016 PRICE VAR [%]' : 'PRICE VAR [%]'}, inplace = True) # Changing 'PRICE VAR' column name
fi_2015['Year'] = 2015 # Creating 'Year' Column

fi_2016.rename(columns = {'Unnamed: 0' : 'Stock', # Changing 'Stock' column name
                          '2017 PRICE VAR [%]' : 'PRICE VAR [%]'}, inplace = True) # Changing 'PRICE VAR' column name
fi_2016['Year'] = 2016 # Creating 'Year' Column

fi_2017.rename(columns = {'Unnamed: 0' : 'Stock', # Changing 'Stock' column name
                          '2018 PRICE VAR [%]' : 'PRICE VAR [%]'}, inplace = True) # Changing 'PRICE VAR' column name
fi_2017['Year'] = 2017 # Creating 'Year' Column

fi_2018.rename(columns = {'Unnamed: 0' : 'Stock', # Changing 'Stock' column name
                          '2019 PRICE VAR [%]' : 'PRICE VAR [%]'}, inplace = True) # Changing 'PRICE VAR' column name
fi_2018['Year'] = 2018 # Creating 'Year' Column

Following the columns changes, we will now check its success.  Success in this case means the column names in all of the datasets have to match.

In [12]:
# The first step to check whether the column names match is to isolate the column names from the datasets.

col_2014 = fi_2014.columns # Isolating the 2014 column names

col_2015 = fi_2015.columns # Isolating the 2015 column names

col_2016 = fi_2016.columns # Isolating the 2016 column names

col_2017 = fi_2017.columns # Isolating the 2017 column names

col_2018 = fi_2018.columns # Isolating the 2018 column names

# The second step is to compare and see if they match.

# For this step using just one of the datasets columns, in this case 2014, to compare with the remaining datasets
# will be enough because if there are no unmatching between 2014 and 2015 and 2014 and 2016, logically speaking
# then the 2015 and the 2016 column names should also match.

print('The number of unequal column names comparing the 2014 dataset with the others:')

print('2014 - 2015:', (col_2014 != col_2015).sum()) # Checking amount of unmatching columns

print('2014 - 2016:', (col_2014 != col_2016).sum()) # Checking amount of unmatching columns

print('2014 - 2017:', (col_2014 != col_2017).sum()) # Checking amount of unmatching columns

print('2014 - 2018:', (col_2014 != col_2018).sum()) # Checking amount of unmatching columns

The number of unequal column names comparing the 2014 dataset with the others:
2014 - 2015: 0
2014 - 2016: 0
2014 - 2017: 0
2014 - 2018: 0


With all of the datasets having the same column names, it is time to merge all of the files.

In [13]:
# To merge all files since they all are dataframes already, let's first first make a list of the dataframes.

data_frames = [fi_2014, fi_2015, fi_2016, fi_2017, fi_2018]

# Now just make them into one by creating a blank dataframe and merging the list created above.

df = pd.DataFrame() # Creating empty dataframe
df = df.append(data_frames) # Merging the files dataframes

In [14]:
# Let's check how the new dataframe looks

df

Stock       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0       PG  7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   
1     VIPS  3.734148e+09          1.1737     2.805625e+09  9.285226e+08   
2       KR  9.837500e+10          0.0182     7.813800e+10  2.023700e+10   
3      RAD  2.552641e+10          0.0053     1.820268e+10  7.323734e+09   
4      GIS  1.790960e+10          0.0076     1.153980e+10  6.369800e+09   
...    ...           ...             ...              ...           ...   
4387  YRIV  0.000000e+00          0.0000     0.000000e+00  0.000000e+00   
4388  YTEN  5.560000e+05         -0.4110     0.000000e+00  5.560000e+05   
4389  ZKIN  5.488438e+07          0.2210     3.659379e+07  1.829059e+07   
4390   ZOM  0.000000e+00          0.0000     0.000000e+00  0.000000e+00   
4391  ZYME  5.301900e+07          0.0243     0.000000e+00  5.301900e+07   

      R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0     0.000000e+00  2.146100e+10        2.146100e+10      1.391000e+10   
1     1.083303e+08  3.441414e+08        7.939267e+08      1.345959e+08   
2     0.000000e+00  1.519600e+10        1.751200e+10      2.725000e+09   
3     0.000000e+00  6.561162e+09        6.586482e+09      7.372520e+08   
4     0.000000e+00  3.474300e+09        3.412400e+09      2.957400e+09   
...            ...           ...                 ...               ...   
4387  0.000000e+00  3.755251e+06        3.755251e+06     -3.755251e+06   
4388  4.759000e+06  5.071000e+06        9.830000e+06     -9.274000e+06   
4389  1.652633e+06  7.020320e+06        8.672953e+06      9.617636e+06   
4390  1.031715e+07  4.521349e+06        1.664863e+07     -1.664863e+07   
4391  5.668400e+07  2.945700e+07        8.614600e+07     -3.312700e+07   

      Interest Expense  ...  Inventory Growth  Asset Growth  \
0         7.090000e+08  ...           -0.0217        0.0359   
1         1.214869e+07  ...               NaN           NaN   
2         4.430000e+08  ...            0.0981        0.1886   
3         4.245910e+08  ...           -0.0510       -0.0189   
4         3.024000e+08  ...            0.0090        0.0215   
...                ...  ...               ...           ...   
4387      1.105849e+07  ...            0.0000       -0.0508   
4388      0.000000e+00  ...            0.0000       -0.2323   
4389      1.239170e+06  ...            0.7706        0.2489   
4390      0.000000e+00  ...            0.0000        0.1568   
4391      1.660000e+05  ...            0.0000        0.8519   

      Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                          0.0316       0.1228              0.0000   
1                             NaN          NaN              1.6484   
2                          0.3268       0.2738              0.0000   
3                          0.1963      -0.0458              0.0000   
4                          0.0274       0.1025              0.0000   
...                           ...          ...                 ...   
4387                      -0.1409      -0.0152              0.0000   
4388                      -0.8602       0.0000              0.0352   
4389                       0.4074      -0.0968              0.2415   
4390                      -0.2200       0.0000              2.7499   
4391                       0.1325       0.0000              0.3577   

      SG&A Expenses Growth              Sector  PRICE VAR [%]  Class  Year  
0                  -0.1746  Consumer Defensive      -9.323276      0  2014  
1                   1.7313  Consumer Defensive     -25.512193      0  2014  
2                   0.0234  Consumer Defensive      33.118297      1  2014  
3                  -0.0060  Consumer Defensive       2.752291      1  2014  
4                  -0.0220  Consumer Defensive      12.897715      1  2014  
...                    ...                 ...            ...    ...   ...  
4387               -0.2602         Real Estate     -90.962099      0  2018  
4

Now that we can see that the new dataframe is built as we expected, it is time to set the index.  For this dataframe the index will be both the 'Year' and the 'Stock' columns.  The year column will keep the separation of the files within the same dataframe, and the stock column is the original index since each stock has difference feature values. 

In [15]:
# Setting the index of the dataframe

df = df.set_index(['Year', 'Stock'])

In [16]:
# Checking the dataframe with the new index

df.head()

Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
Year Stock                                                                
2014 PG     7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   
     VIPS   3.734148e+09          1.1737     2.805625e+09  9.285226e+08   
     KR     9.837500e+10          0.0182     7.813800e+10  2.023700e+10   
     RAD    2.552641e+10          0.0053     1.820268e+10  7.323734e+09   
     GIS    1.790960e+10          0.0076     1.153980e+10  6.369800e+09   

            R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
Year Stock                                                                     
2014 PG     0.000000e+00  2.146100e+10        2.146100e+10      1.391000e+10   
     VIPS   1.083303e+08  3.441414e+08        7.939267e+08      1.345959e+08   
     KR     0.000000e+00  1.519600e+10        1.751200e+10      2.725000e+09   
     RAD    0.000000e+00  6.561162e+09        6.586482e+09      7.372520e+08   
     GIS    0.000000e+00  3.474300e+09        3.412400e+09      2.957400e+09   

            Interest Expense  Earnings before Tax  ...  Receivables growth  \
Year Stock                                         ...                       
2014 PG         7.090000e+08         1.449400e+10  ...             -0.0187   
     VIPS       1.214869e+07         1.753823e+08  ...                 NaN   
     KR         4.430000e+08         2.270000e+09  ...              0.0618   
     RAD        4.245910e+08         2.502180e+08  ...              0.0211   
     GIS        3.024000e+08         2.707700e+09  ...              0.0257   

            Inventory Growth  Asset Growth  Book Value per Share Growth  \
Year Stock                                                                
2014 PG              -0.0217        0.0359                       0.0316   
     VIPS                NaN           NaN                          NaN   
     KR               0.0981        0.1886                       0.3268   
     RAD             -0.0510       -0.0189                       0.1963   
     GIS              0.0090        0.0215                       0.0274   

            Debt Growth  R&D Expense Growth  SG&A Expenses Growth  \
Year Stock                                                          
2014 PG          0.1228              0.0000               -0.1746   
     VIPS           NaN              1.6484                1.7313   
     KR          0.2738              0.0000                0.0234   
     RAD        -0.0458              0.0000               -0.0060   
     GIS         0.1025              0.0000               -0.0220   

                        Sector  PRICE VAR [%]  Class  
Year Stock                                            
2014 PG     Consumer Defensive      -9.323276      0  
     VIPS   Consumer Defensive     -25.512193      0  
     KR     Consumer Defensive      33.118297      1  
     RAD    Consumer Defensive       2.752291      1  
     GIS    Consumer Defensive      12.897715      1  

[5 rows x 224 columns]

I will now begin cleaning the dataframe.  The first thing that I will do is to eliminate the stocks that have more than 75% of their values missing.  To do that, I will first transpose the dataframe so that I can get the index (stocks and year) as column names.

In [17]:
df_T = df.T

df_T

Year                                2014                      \
Stock                                 PG                VIPS   
Revenue                       7.4401e+10         3.73415e+09   
Revenue Growth                   -0.0713              1.1737   
Cost of Revenue                3.903e+10         2.80563e+09   
Gross Profit                  3.5371e+10         9.28523e+08   
R&D Expenses                           0          1.0833e+08   
...                                  ...                 ...   
R&D Expense Growth                     0              1.6484   
SG&A Expenses Growth             -0.1746              1.7313   
Sector                Consumer Defensive  Consumer Defensive   
PRICE VAR [%]                   -9.32328            -25.5122   
Class                                  0                   0   

Year                                                          \
Stock                                 KR                 RAD   
Revenue                       9.8375e+10         2.55264e+10   
Revenue Growth                    0.0182              0.0053   
Cost of Revenue               7.8138e+10         1.82027e+10   
Gross Profit                  2.0237e+10         7.32373e+09   
R&D Expenses                           0                   0   
...                                  ...                 ...   
R&D Expense Growth                     0                   0   
SG&A Expenses Growth              0.0234              -0.006   
Sector                Consumer Defensive  Consumer Defensive   
PRICE VAR [%]                    33.1183             2.75229   
Class                                  1                   1   

Year                                                          \
Stock                                GIS                  PM   
Revenue                      1.79096e+10          8.0106e+10   
Revenue Growth                    0.0076               0.001   
Cost of Revenue              1.15398e+10          6.0775e+10   
Gross Profit                  6.3698e+09          1.9331e+10   
R&D Expenses                           0                   0   
...                                  ...                 ...   
R&D Expense Growth                     0                   0   
SG&A Expenses Growth              -0.022              0.0161   
Sector                Consumer Defensive  Consumer Defensive   
PRICE VAR [%]                    12.8977             13.9809   
Class                                  1                   1   

Year                                                          \
Stock                                 KO                 WMT   
Revenue                       4.5998e+10         4.76294e+11   
Revenue Growth                   -0.0183              0.0163   
Cost of Revenue               1.7889e+10         3.58069e+11   
Gross Profit                  2.8109e+10         1.18225e+11   
R&D Expenses                           0                   0   
...                                  ...                 ...   
R&D Expense Growth                     0                   0   
SG&A Expenses Growth             -0.0053              0.0307   
Sector                Consumer Defensive  Consumer Defensive   
PRICE VAR [%]                    5.33941            -26.6537   
Class                                  1                   0   

Year                                                          ...  \
Stock                               MDLZ                  MO  ...   
Revenue                       3.4244e+10          2.4522e+10  ...   
Revenue Growth                   -0.0299              0.0023  ...   
Cost of Revenue               2.1647e+10          1.4362e+10  ...   
Gross Profit                  1.2597e+10           1.016e+10  ...   
R&D Expenses                           0                   0  ...   
...                                  ...                 ...  ...   
R&D Expense Growth                     0                   0  ...   
SG&A Expenses Growth             -0.0256       

Now I will collect the stocks that have 75% or more missing values.

In [18]:
missing_75 = df_T.loc[:,df_T.isna().mean() >= 0.75].columns

missing_75

MultiIndex([(2014,  'TWNK'),
            (2014,  'TEDU'),
            (2014,  'ADAP'),
            (2014,  'IMUC'),
            (2014,  'ANTH'),
            (2014, 'CDMOP'),
            (2014, 'CYCCP'),
            (2014,   'HUD'),
            (2014,   'CCZ'),
            (2014, 'HOVNP'),
            ...
            (2018,   'SCD'),
            (2018, 'SLMBP'),
            (2018,  'SSBI'),
            (2018, 'TCBIP'),
            (2018,   'TDI'),
            (2018,   'TDJ'),
            (2018,   'UBP'),
            (2018,   'VCF'),
            (2018, 'VRTSP'),
            (2018, 'WHLRD')],
           names=['Year', 'Stock'], length=1784)

We can see that less than 10% of the stocks has 75% or more missing data in them.  Those stocks will then be extracted from the dataframe because they may be the cause of noise in later stages of the project and there are relatively very few of them.

In [19]:
df.drop(index = missing_75, inplace = True)

df.shape

(20293, 224)

After eliminating stocks with little to no useful information, it is time to begin working on the features.  The first step into checking the features will be to check the average missing values.

In [20]:
mv = pd.DataFrame(df.isna().mean(), columns = ['Average Missing Values']) # Creates a Missing Values DataFrame

mv.sort_values(by = ['Average Missing Values'], inplace = True) # Sorts the DataFrame

mv

Average Missing Values
Class                                                     0.000000
Sector                                                    0.000000
PRICE VAR [%]                                             0.000000
Gross Profit                                              0.000000
Financing Cash Flow                                       0.000148
...                                                            ...
10Y Revenue Growth (per Share)                            0.381807
10Y Shareholders Equity Growth (per Share)                0.389100
shortTermCoverageRatios                                   0.441827
cashConversionCycle                                       0.999064
operatingCycle                                            0.999064

[224 rows x 1 columns]

I notice that 2 features (cashConversionCycle and operatingCycle) have over 99% of their data missing.  Those features are deemed useless and will be droped in the next step.

In [21]:
df.drop(columns = ['cashConversionCycle', 'operatingCycle'], inplace = True)

In [22]:
df.shape

(20293, 222)

The next step will be to "filter" out the features that have duplicate values and/or names. The first part of this "filtering" stage is to extract the features with duplicate values.  To do that I will tranpose the dataframe and check for stocks with the least amount of missing values.

In [23]:
sorted(df.T.isna().mean())

[0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0

Perfect!

There are now plenty of stocks that have absolutely 0 missing values!

Those stocks will now be used to collect the features that have duplicate values.  For that I will create a sample dataframe that contains only the stocks that has zero missing values, transpose it, and use the duplicated() method.

In [24]:
df_sample = df.T.loc[:,df.T.isna().mean() == 0.0]

In [25]:
df_sample.drop(index = ['Class', 'Sector'], inplace = True)

In [26]:
df_sample

Year                               2014                                     \
Stock                                KO         NWL        CPB         PEP   
Revenue                      4.5998e+10   5.727e+09  8.268e+09  6.6683e+10   
Revenue Growth                  -0.0183      0.0214     0.0268       0.004   
Cost of Revenue              1.7889e+10  3.5236e+09  5.297e+09  3.1238e+10   
Gross Profit                 2.8109e+10  2.2034e+09  2.971e+09  3.5445e+10   
R&D Expenses                          0           0   1.22e+08           0   
...                                 ...         ...        ...         ...   
Book Value per Share Growth     -0.0763     -0.0658     0.3269     -0.2646   
Debt Growth                      0.1258      0.3515    -0.0984      -0.025   
R&D Expense Growth                    0           0    -0.0469           0   
SG&A Expenses Growth            -0.0053      0.0579    -0.0733      0.0164   
PRICE VAR [%]                   5.33941     18.6038    22.6288      8.8867   

Year                                                                          \
Stock                               ADM         TAP          SYY         CHD   
Revenue                      8.1201e+10  4.1463e+09  4.65167e+10  3.2976e+09   
Revenue Growth                  -0.0958     -0.0142       0.0474      0.0323   
Cost of Revenue              7.6433e+10  2.4933e+09  3.83357e+10  1.8447e+09   
Gross Profit                  4.768e+09   1.653e+09  8.18104e+09  1.4529e+09   
R&D Expenses                          0           0            0           0   
...                                 ...         ...          ...         ...   
Book Value per Share Growth      -0.016     -0.0956       0.0203     -0.0624   
Debt Growth                     -0.1761     -0.1657      -0.0539      0.3527   
R&D Expense Growth                    0           0            0           0   
SG&A Expenses Growth             0.0989      -0.025       0.0404      -0.005   
PRICE VAR [%]                  -27.7281     29.9029      5.34612     10.7926   

Year                                                   ...         2018  \
Stock                                 EL          PPC  ...          SGA   
Revenue                      1.09688e+10  8.58336e+09  ...  1.24829e+08   
Revenue Growth                    0.0773       0.0205  ...       0.0565   
Cost of Revenue               2.1582e+09  7.18937e+09  ...   9.3727e+07   
Gross Profit                  8.8106e+09    1.394e+09  ...   3.1102e+07   
R&D Expenses                           0            0  ...            0   
...                                  ...          ...  ...          ...   
Book Value per Share Growth       0.1771       0.4717  ...       0.0263   
Debt Growth                      -0.0009      -0.9953  ...         -0.2   
R&D Expense Growth                     0            0  ...            0   
SG&A Expenses Growth               0.059       0.0424  ...      -0.0256   
PRICE VAR [%]                    17.9308     -17.3446  ...     -5.62535   

Year                                                                          \
Stock                               SSNT         SVT        TGC         TRNS   
Revenue                      4.10003e+07  4.7857e+07  5.871e+06  1.55141e+08   
Revenue Growth                    0.1764      0.1547     0.2537       0.0781   
Cost of Revenue              2.40799e+07  3.5772e+07  3.591e+06    1.177e+08   
Gross Profit                 1.69204e+07  1.2085e+07   2.28e+06   3.7441e+07   
R&D Expenses                           0           0          0            0   
...                                  ...         ...        ...          ...   
Book Value per Share Growth       0.0223      0.0878     0.2417       0.1616   
Debt Growth                        1.603     -0.1372     0.3778      -0.1634   
R&D Expense Growth                     0           0          0            0   
SG&A Expenses Growth               0.289      -0.013     0.0632        0.051   
PRICE VA

In [27]:
duplicated_values = df_sample.loc[df_sample.duplicated(),:]

In [28]:
duplicated_values

Year                               2014                                 \
Stock                                KO        NWL       CPB       PEP   
nIperEBT                       0.763308   0.809167  0.698387   0.74759   
eBTperEBIT                     0.950624   0.885454  0.910426  0.905519   
eBITperRevenue                 0.212661  0.0920726  0.164731   0.14428   
cashFlowCoverageRatios         0.254282    0.25549   0.22391  0.363567   
Operating Cash Flow per Share    2.4196     2.2966    2.8631    6.9622   
Free Cash Flow per Share          1.922       1.71     2.191     5.144   
Cash per Share                   4.1053     0.7222    0.7389    4.0649   
PE ratio                        26.0617    27.8029   15.3007   22.4432   
Price to Sales Ratio             4.0204     1.8031    1.6034     2.171   
POCF ratio                      17.4214    16.2848   14.7459   13.7794   
PFCF ratio                      21.9662    22.2716   23.1014   18.8051   
PB ratio                         6.0879     5.6125    8.2107    8.2378   
PTB ratio                       46.7543          0         0         0   
Debt to Equity                   1.3768     1.3406    2.4861    1.6439   
Debt to Assets                   0.4536     0.3781    0.4949    0.4098   
Current ratio                     1.019      1.214     0.703     1.142   
Interest Coverage               20.2526     8.7301   11.1639   10.5842   
Payout Ratio                      0.753      0.482     0.452     0.588   
ROIC                             0.1004     0.0776    0.1186    0.1227   
Return on Tangible Assets        0.4656     0.3924     0.606    0.3287   
Days Sales Outstanding          35.4383    79.5518   29.5779   36.4053   
Days Payables Outstanding        73.273    42.9625    23.265   71.2451   
Days of Inventory on Hand      -63.2512   -73.3916  -70.0094  -36.7243   
Payables Turnover                1.1559     2.4753    3.5276    1.5615   
Inventory Turnover              14.4262     8.2231    8.5193    20.355   
ROE                              0.2341     0.2041    0.5362    0.3705   

Year                                                                     \
Stock                                ADM       TAP        SYY       CHD   
nIperEBT                         0.71936  0.881647   0.631281  0.662346   
eBTperEBIT                      0.902657  0.800824   0.922631  0.957995   
eBITperRevenue                 0.0426349  0.175578  0.0343826  0.197811   
cashFlowCoverageRatios          0.873322  0.406239   0.546202  0.497239   
Operating Cash Flow per Share     7.5697    6.9654     2.5475    1.9996   
Free Cash Flow per Share          6.2006      5.61      1.723     1.739   
Cash per Share                     1.683     3.378      0.953    1.5655   
PE ratio                         15.1163   26.8058    23.5535   25.7549   
Price to Sales Ratio              0.4123    2.9518     0.4706    3.1971   
POCF ratio                        6.7726    9.5032     14.665   19.5128   
PFCF ratio                        8.3863   13.2846    22.0467   22.6633   
PB ratio                          1.7139     1.758     4.1708    5.0205   
PTB ratio                         2.0719         0     6.9983         0   
Debt to Equity                    0.2883    0.4032     0.5189     0.517   
Debt to Assets                    0.1286    0.2268      0.208    0.2493   
Current ratio                      1.668     0.678       1.53     1.125   
Interest Coverage                 10.273    5.0207    12.9251   23.8066   
Payout Ratio                       0.279     0.532      0.717     0.405   
ROIC                              0.0774    0.0087     0.1155    0.1169   
Return on Tangible Assets         0.1109    0.4894     0.1479    0.5678   
Days Sales Outstanding           12.1545   54.7284    26.6685   35.7407   
Days Payables Outstanding        45.8492   114.879    22.2141   56.1956   
Days of Inventory on Hand       -44.7648  -33.1139   -24.7742  -48.6548   
Payables Turnover                  2.138    0.7122  

The dataframe above tells me which are the duplicate features.  For my next steps I will create a list of those features and use it to drop from the main dataframe.

In [29]:
duplicated_values_columns = list(duplicated_values.index)
duplicated_values_columns

['nIperEBT',
 'eBTperEBIT',
 'eBITperRevenue',
 'cashFlowCoverageRatios',
 'Operating Cash Flow per Share',
 'Free Cash Flow per Share',
 'Cash per Share',
 'PE ratio',
 'Price to Sales Ratio',
 'POCF ratio',
 'PFCF ratio',
 'PB ratio',
 'PTB ratio',
 'Debt to Equity',
 'Debt to Assets',
 'Current ratio',
 'Interest Coverage',
 'Payout Ratio',
 'ROIC',
 'Return on Tangible Assets',
 'Days Sales Outstanding',
 'Days Payables Outstanding',
 'Days of Inventory on Hand',
 'Payables Turnover',
 'Inventory Turnover',
 'ROE']

In [30]:
df.drop(columns = duplicated_values_columns, inplace = True)

In [31]:
df

Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
Year Stock                                                                
2014 PG     7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   
     VIPS   3.734148e+09          1.1737     2.805625e+09  9.285226e+08   
     KR     9.837500e+10          0.0182     7.813800e+10  2.023700e+10   
     RAD    2.552641e+10          0.0053     1.820268e+10  7.323734e+09   
     GIS    1.790960e+10          0.0076     1.153980e+10  6.369800e+09   
...                  ...             ...              ...           ...   
2018 YRIV   0.000000e+00          0.0000     0.000000e+00  0.000000e+00   
     YTEN   5.560000e+05         -0.4110     0.000000e+00  5.560000e+05   
     ZKIN   5.488438e+07          0.2210     3.659379e+07  1.829059e+07   
     ZOM    0.000000e+00          0.0000     0.000000e+00  0.000000e+00   
     ZYME   5.301900e+07          0.0243     0.000000e+00  5.301900e+07   

            R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
Year Stock                                                                     
2014 PG     0.000000e+00  2.146100e+10        2.146100e+10      1.391000e+10   
     VIPS   1.083303e+08  3.441414e+08        7.939267e+08      1.345959e+08   
     KR     0.000000e+00  1.519600e+10        1.751200e+10      2.725000e+09   
     RAD    0.000000e+00  6.561162e+09        6.586482e+09      7.372520e+08   
     GIS    0.000000e+00  3.474300e+09        3.412400e+09      2.957400e+09   
...                  ...           ...                 ...               ...   
2018 YRIV   0.000000e+00  3.755251e+06        3.755251e+06     -3.755251e+06   
     YTEN   4.759000e+06  5.071000e+06        9.830000e+06     -9.274000e+06   
     ZKIN   1.652633e+06  7.020320e+06        8.672953e+06      9.617636e+06   
     ZOM    1.031715e+07  4.521349e+06        1.664863e+07     -1.664863e+07   
     ZYME   5.668400e+07  2.945700e+07        8.614600e+07     -3.312700e+07   

            Interest Expense  Earnings before Tax  ...  Receivables growth  \
Year Stock                                         ...                       
2014 PG         7.090000e+08         1.449400e+10  ...             -0.0187   
     VIPS       1.214869e+07         1.753823e+08  ...                 NaN   
     KR         4.430000e+08         2.270000e+09  ...              0.0618   
     RAD        4.245910e+08         2.502180e+08  ...              0.0211   
     GIS        3.024000e+08         2.707700e+09  ...              0.0257   
...                      ...                  ...  ...                 ...   
2018 YRIV       1.105849e+07        -1.482451e+07  ...              0.0000   
     YTEN       0.000000e+00        -9.170000e+06  ...              0.3445   
     ZKIN       1.239170e+06         8.416324e+06  ...              0.1605   
     ZOM        0.000000e+00        -1.664769e+07  ...              0.8980   
     ZYME       1.660000e+05        -3.438500e+07  ...             -0.4185   

            Inventory Growth  Asset Growth  Book Value per Share Growth  \
Year Stock                                                                
2014 PG              -0.0217        0.0359                       0.0316   
     VIPS                NaN           NaN                          NaN   
     KR               0.0981        0.1886                       0.3268   
     RAD             -0.0510       -0.0189                       0.1963   
     GIS              0.0090        0.0215                       0.0274   
...                      ...           ...                          ...   
2018 YRIV             0.0000       -0.0508                      -0.1409   
     YTEN             0.0000       -0.2323                      -0.8602   
     ZKIN             0.7706        0.2489                       0.4074   
     ZOM              0.0000        0.1568                      -0.2200   
     ZYME             0.0000        0.8519                       0.1325   

         

A greate progress has beem made so far.  Now I will see if there are any features that have duplicate names.  These columns may not have been filtered when looking for duplicate values because some of them may have more missing values than their duplicates.  For this step I will find the feature names that are duplicates and extract the ones that have the most missing values.

To catch duplicate feature names, I will create a list of the dataframe column names lower-cased and no space.

In [32]:
lst1 = [] # Creating an empty list.

for c in df.columns:
    lst1.append(c.replace(' ', '').lower()) # Adding the column names to the list lower-cased and without space.
    
# Let's check the newly created list

lst1


['revenue',
 'revenuegrowth',
 'costofrevenue',
 'grossprofit',
 'r&dexpenses',
 'sg&aexpense',
 'operatingexpenses',
 'operatingincome',
 'interestexpense',
 'earningsbeforetax',
 'incometaxexpense',
 'netincome-non-controllingint',
 'netincome-discontinuedops',
 'netincome',
 'preferreddividends',
 'netincomecom',
 'eps',
 'epsdiluted',
 'weightedaverageshsout',
 'weightedaverageshsout(dil)',
 'dividendpershare',
 'grossmargin',
 'ebitdamargin',
 'ebitmargin',
 'profitmargin',
 'freecashflowmargin',
 'ebitda',
 'ebit',
 'consolidatedincome',
 'earningsbeforetaxmargin',
 'netprofitmargin',
 'cashandcashequivalents',
 'short-terminvestments',
 'cashandshort-terminvestments',
 'receivables',
 'inventories',
 'totalcurrentassets',
 'property,plant&equipmentnet',
 'goodwillandintangibleassets',
 'long-terminvestments',
 'taxassets',
 'totalnon-currentassets',
 'totalassets',
 'payables',
 'short-termdebt',
 'totalcurrentliabilities',
 'long-termdebt',
 'totaldebt',
 'deferredrevenue',
 't

The list is now created and the order of the features in the list is the exact same order they appear in the original dataframe.  To keep this order, I will create a new dataframe out of that list because the index of this new dataframe will be the same index of the column in the original dataframe and by building a dataframe I can use the duplicated() method once again to check for duplicated names.

In [33]:
# Creating the columns names dataframe

col_df = pd.DataFrame({"Column Names" : lst1})

# Checking the column names dataframe

col_df

Column Names
0               revenue
1         revenuegrowth
2         costofrevenue
3           grossprofit
4           r&dexpenses
..                  ...
191    r&dexpensegrowth
192  sg&aexpensesgrowth
193              sector
194         pricevar[%]
195               class

[196 rows x 1 columns]

In [34]:
# Creating the duplicate dataframe with the 'second' names.

duplicates_last = col_df[col_df.duplicated()]

# Creating the duplicate dataframe with the 'first' names.

duplicates_first = col_df[col_df.duplicated(keep = 'last')]

In [35]:
duplicates_last

Column Names
93   netprofitmargin
141    dividendyield

In [36]:
duplicates_first

Column Names
30  netprofitmargin
84    dividendyield

There are 2 columns that have the same name but different column index.  To find those columns in the original dataframe I will use the index of the new dataframes.

While fiding the real name, I will also check and see which one of them have the least amount of missing data.  The column with the most missing data will be dropped.

In [37]:
duplicate_column_names = {
                          93 : [df.iloc[:,93].name, df.iloc[:,93].isna().sum()],
                          30 : [df.iloc[:,30].name, df.iloc[:,30].isna().sum()],
                          84 : [df.iloc[:,84].name, df.iloc[:,84].isna().sum()],
                          141 : [df.iloc[:,141].name, df.iloc[:,141].isna().sum()]
                         }
duplicate_column_names = pd.DataFrame.from_dict(duplicate_column_names, orient = 'index',
                                                columns = ['Column Name', 'Missing Values'])



In [38]:
duplicate_column_names.index.name = 'Original Index'

In [39]:
duplicate_column_names

Column Name  Missing Values
Original Index                                   
93                netProfitMargin            1378
30              Net Profit Margin             169
84                  dividendYield            1514
141                Dividend Yield             466

The features at index 93 and 84 will be the ones extracted because they have the most missing values then their respective duplicates.

In [40]:
df.drop(columns = [df.iloc[:,93].name, df.iloc[:,84].name], inplace = True)

In [41]:
df

Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
Year Stock                                                                
2014 PG     7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   
     VIPS   3.734148e+09          1.1737     2.805625e+09  9.285226e+08   
     KR     9.837500e+10          0.0182     7.813800e+10  2.023700e+10   
     RAD    2.552641e+10          0.0053     1.820268e+10  7.323734e+09   
     GIS    1.790960e+10          0.0076     1.153980e+10  6.369800e+09   
...                  ...             ...              ...           ...   
2018 YRIV   0.000000e+00          0.0000     0.000000e+00  0.000000e+00   
     YTEN   5.560000e+05         -0.4110     0.000000e+00  5.560000e+05   
     ZKIN   5.488438e+07          0.2210     3.659379e+07  1.829059e+07   
     ZOM    0.000000e+00          0.0000     0.000000e+00  0.000000e+00   
     ZYME   5.301900e+07          0.0243     0.000000e+00  5.301900e+07   

            R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
Year Stock                                                                     
2014 PG     0.000000e+00  2.146100e+10        2.146100e+10      1.391000e+10   
     VIPS   1.083303e+08  3.441414e+08        7.939267e+08      1.345959e+08   
     KR     0.000000e+00  1.519600e+10        1.751200e+10      2.725000e+09   
     RAD    0.000000e+00  6.561162e+09        6.586482e+09      7.372520e+08   
     GIS    0.000000e+00  3.474300e+09        3.412400e+09      2.957400e+09   
...                  ...           ...                 ...               ...   
2018 YRIV   0.000000e+00  3.755251e+06        3.755251e+06     -3.755251e+06   
     YTEN   4.759000e+06  5.071000e+06        9.830000e+06     -9.274000e+06   
     ZKIN   1.652633e+06  7.020320e+06        8.672953e+06      9.617636e+06   
     ZOM    1.031715e+07  4.521349e+06        1.664863e+07     -1.664863e+07   
     ZYME   5.668400e+07  2.945700e+07        8.614600e+07     -3.312700e+07   

            Interest Expense  Earnings before Tax  ...  Receivables growth  \
Year Stock                                         ...                       
2014 PG         7.090000e+08         1.449400e+10  ...             -0.0187   
     VIPS       1.214869e+07         1.753823e+08  ...                 NaN   
     KR         4.430000e+08         2.270000e+09  ...              0.0618   
     RAD        4.245910e+08         2.502180e+08  ...              0.0211   
     GIS        3.024000e+08         2.707700e+09  ...              0.0257   
...                      ...                  ...  ...                 ...   
2018 YRIV       1.105849e+07        -1.482451e+07  ...              0.0000   
     YTEN       0.000000e+00        -9.170000e+06  ...              0.3445   
     ZKIN       1.239170e+06         8.416324e+06  ...              0.1605   
     ZOM        0.000000e+00        -1.664769e+07  ...              0.8980   
     ZYME       1.660000e+05        -3.438500e+07  ...             -0.4185   

            Inventory Growth  Asset Growth  Book Value per Share Growth  \
Year Stock                                                                
2014 PG              -0.0217        0.0359                       0.0316   
     VIPS                NaN           NaN                          NaN   
     KR               0.0981        0.1886                       0.3268   
     RAD             -0.0510       -0.0189                       0.1963   
     GIS              0.0090        0.0215                       0.0274   
...                      ...           ...                          ...   
2018 YRIV             0.0000       -0.0508                      -0.1409   
     YTEN             0.0000       -0.2323                      -0.8602   
     ZKIN             0.7706        0.2489                       0.4074   
     ZOM              0.0000        0.1568                      -0.2200   
     ZYME             0.0000        0.8519                       0.1325   

         

Now that I have taken care of all of the columns that were duplicates, the remaining problem is the amount of missing values.  For that I will check the average missing values per columns.  I have learned earlier that very few of the columns have no missing values.  Since I will be picking from the remaining columns to use in the next phases of the project and to a good array of options, I cannot just keep the features that have 0 missing values.

To try and find out the columns I will keep to choose from, I will first check the average missing values in each of them.

In [42]:
sorted(df.isna().mean())

[0.0,
 0.0,
 0.0,
 0.0,
 0.00014783422855171733,
 0.00014783422855171733,
 0.00034494653328734045,
 0.0006406149903907751,
 0.0006898930665746809,
 0.0012812299807815502,
 0.001330508056965456,
 0.001330508056965456,
 0.001330508056965456,
 0.001330508056965456,
 0.0014783422855171734,
 0.0018725668949884197,
 0.0021189572759079487,
 0.0023653476568274772,
 0.0030059626472182526,
 0.0034494653328734047,
 0.0036958557137929337,
 0.0036958557137929337,
 0.0036958557137929337,
 0.004237914551815897,
 0.004237914551815897,
 0.004237914551815897,
 0.004385748780367614,
 0.004484304932735426,
 0.004878529542206672,
 0.005223476075494012,
 0.005617700684965259,
 0.005617700684965259,
 0.006011925294436505,
 0.006061203370620411,
 0.006110481446804317,
 0.008327994875080076,
 0.008327994875080076,
 0.008327994875080076,
 0.008426551027447889,
 0.008475829103631795,
 0.008771497560735229,
 0.008919331789286946,
 0.009067166017838664,
 0.009215000246390381,
 0.009264278322574288,
 0.009658502932

By doing a quick glance, I can see that most columns has up to 5% of missing data.  Just to be sure, I will check how many columns has up to 5% of missing values.

In [43]:
(df.isna().mean() <= 0.05).sum()

126

126 features gives me a great number of options from which to pick!  I will now keep only those features.

In [44]:
df = df.loc[:,df.isna().mean() <= 0.05]

In [45]:
df

Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
Year Stock                                                                
2014 PG     7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   
     VIPS   3.734148e+09          1.1737     2.805625e+09  9.285226e+08   
     KR     9.837500e+10          0.0182     7.813800e+10  2.023700e+10   
     RAD    2.552641e+10          0.0053     1.820268e+10  7.323734e+09   
     GIS    1.790960e+10          0.0076     1.153980e+10  6.369800e+09   
...                  ...             ...              ...           ...   
2018 YRIV   0.000000e+00          0.0000     0.000000e+00  0.000000e+00   
     YTEN   5.560000e+05         -0.4110     0.000000e+00  5.560000e+05   
     ZKIN   5.488438e+07          0.2210     3.659379e+07  1.829059e+07   
     ZOM    0.000000e+00          0.0000     0.000000e+00  0.000000e+00   
     ZYME   5.301900e+07          0.0243     0.000000e+00  5.301900e+07   

            R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
Year Stock                                                                     
2014 PG     0.000000e+00  2.146100e+10        2.146100e+10      1.391000e+10   
     VIPS   1.083303e+08  3.441414e+08        7.939267e+08      1.345959e+08   
     KR     0.000000e+00  1.519600e+10        1.751200e+10      2.725000e+09   
     RAD    0.000000e+00  6.561162e+09        6.586482e+09      7.372520e+08   
     GIS    0.000000e+00  3.474300e+09        3.412400e+09      2.957400e+09   
...                  ...           ...                 ...               ...   
2018 YRIV   0.000000e+00  3.755251e+06        3.755251e+06     -3.755251e+06   
     YTEN   4.759000e+06  5.071000e+06        9.830000e+06     -9.274000e+06   
     ZKIN   1.652633e+06  7.020320e+06        8.672953e+06      9.617636e+06   
     ZOM    1.031715e+07  4.521349e+06        1.664863e+07     -1.664863e+07   
     ZYME   5.668400e+07  2.945700e+07        8.614600e+07     -3.312700e+07   

            Interest Expense  Earnings before Tax  ...  Free Cash Flow growth  \
Year Stock                                         ...                          
2014 PG         7.090000e+08         1.449400e+10  ...                -0.0695   
     VIPS       1.214869e+07         1.753823e+08  ...                -0.3626   
     KR         4.430000e+08         2.270000e+09  ...                 0.3935   
     RAD        4.245910e+08         2.502180e+08  ...                -0.2691   
     GIS        3.024000e+08         2.707700e+09  ...                -0.1936   
...                      ...                  ...  ...                    ...   
2018 YRIV       1.105849e+07        -1.482451e+07  ...                -2.0605   
     YTEN       0.000000e+00        -9.170000e+06  ...                -0.0716   
     ZKIN       1.239170e+06         8.416324e+06  ...                -3.7263   
     ZOM        0.000000e+00        -1.664769e+07  ...                -0.6197   
     ZYME       1.660000e+05        -3.438500e+07  ...                14.0145   

            Receivables growth  Inventory Growth  Asset Growth  Debt Growth  \
Year Stock                                                                    
2014 PG                -0.0187           -0.0217        0.0359       0.1228   
     VIPS                  NaN               NaN           NaN          NaN   
     KR                 0.0618            0.0981        0.1886       0.2738   
     RAD                0.0211           -0.0510       -0.0189      -0.0458   
     GIS                0.0257            0.0090        0.0215       0.1025   
...                        ...               ...           ...          ...   
2018 YRIV               0.0000            0.0000       -0.0508      -0.0152   
     YTEN               0.3445            0.0000       -0.2323       0.0000   
     ZKIN               0.1605            0.7706        0.2489      -0.0968   
     ZOM                0.8980            0.0000        0.1568       0.0000  

To take care of the remaining missing values, I will check how many stocks I can keep if I only kept the ones that have no missing values.

In [46]:
(df.T.isna().sum() == 0).sum()

17314

After the changes that I have made so far, if I were to eliminate all of the missing values by keeping just the stocks that do not have any missing values, I would be keeping 17314 stocks, which is a very good amount to keep for the next phases of the project.

In [47]:
df = df.loc[df.T.isna().sum() == 0, :]

In [48]:
df

Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
Year Stock                                                                
2014 PG     7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   
     KR     9.837500e+10          0.0182     7.813800e+10  2.023700e+10   
     RAD    2.552641e+10          0.0053     1.820268e+10  7.323734e+09   
     GIS    1.790960e+10          0.0076     1.153980e+10  6.369800e+09   
     PM     8.010600e+10          0.0010     6.077500e+10  1.933100e+10   
...                  ...             ...              ...           ...   
2018 XBIO   0.000000e+00         -1.0000     0.000000e+00  0.000000e+00   
     XELB   3.546600e+07          0.1186     2.702000e+06  3.276400e+07   
     YTEN   5.560000e+05         -0.4110     0.000000e+00  5.560000e+05   
     ZKIN   5.488438e+07          0.2210     3.659379e+07  1.829059e+07   
     ZYME   5.301900e+07          0.0243     0.000000e+00  5.301900e+07   

            R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
Year Stock                                                                     
2014 PG              0.0  2.146100e+10        2.146100e+10      1.391000e+10   
     KR              0.0  1.519600e+10        1.751200e+10      2.725000e+09   
     RAD             0.0  6.561162e+09        6.586482e+09      7.372520e+08   
     GIS             0.0  3.474300e+09        3.412400e+09      2.957400e+09   
     PM              0.0  7.001000e+09        7.629000e+09      1.170200e+10   
...                  ...           ...                 ...               ...   
2018 XBIO      2883952.0  4.392375e+06        7.276327e+06     -7.276327e+06   
     XELB            0.0  2.446700e+07        2.883400e+07      3.930000e+06   
     YTEN      4759000.0  5.071000e+06        9.830000e+06     -9.274000e+06   
     ZKIN      1652633.0  7.020320e+06        8.672953e+06      9.617636e+06   
     ZYME     56684000.0  2.945700e+07        8.614600e+07     -3.312700e+07   

            Interest Expense  Earnings before Tax  ...  Free Cash Flow growth  \
Year Stock                                         ...                          
2014 PG         7.090000e+08         1.449400e+10  ...                -0.0695   
     KR         4.430000e+08         2.270000e+09  ...                 0.3935   
     RAD        4.245910e+08         2.502180e+08  ...                -0.2691   
     GIS        3.024000e+08         2.707700e+09  ...                -0.1936   
     PM         1.052000e+09         1.059000e+10  ...                -0.2629   
...                      ...                  ...  ...                    ...   
2018 XBIO      -5.090000e+02        -7.300458e+06  ...                -5.3379   
     XELB       1.011000e+06         2.919000e+06  ...                 0.1450   
     YTEN       0.000000e+00        -9.170000e+06  ...                -0.0716   
     ZKIN       1.239170e+06         8.416324e+06  ...                -3.7263   
     ZYME       1.660000e+05        -3.438500e+07  ...                14.0145   

            Receivables growth  Inventory Growth  Asset Growth  Debt Growth  \
Year Stock                                                                    
2014 PG                -0.0187           -0.0217        0.0359       0.1228   
     KR                 0.0618            0.0981        0.1886       0.2738   
     RAD                0.0211           -0.0510       -0.0189      -0.0458   
     GIS                0.0257            0.0090        0.0215       0.1025   
     PM                 0.0392           -0.1274       -0.0781       0.0642   
...                        ...               ...           ...          ...   
2018 XBIO               0.0000            0.0000       -0.2470       0.0000   
     XELB               0.2910            0.0000        0.0225      -0.2154   
     YTEN               0.3445            0.0000       -0.2323       0.0000   
     ZKIN               0.1605            0.7706        0.2489      -0.0968  